rfcs/0000-poa-testnet/0000-poa-testnet.md

```diff
@@ -0,0 +1,124 @@
+
+## attest a new block
+
+let's pre assuming the validator list already exists. Our protocol is simple enough that the change of validator list will not affect our purpose, we can assume the validator list is fixed for now.
+
+Validators use a round-robin style to attest blocks.
+For block `n`, the validator which index equals `n % VALIDATOR_COUNT` suppose to be the attester. However, a validator can fail to produce a block in time due to network congestion or other reasons, in this case, other validators should produce the block `n`.
+
+Validators can use a simple strategy to achieve this:
+
+* If `INDEX == n % VALIDATOR_COUNT`, wait for `BLOCK_INTERVAL` seconds then produce a new block with difficulty set to `2`.
+* If `INDEX != n % VALIDATOR_COUNT`, wait for `BLOCK_INTERVAL + rand(VALIDATOR_COUNT) * 0.5` seconds then produce a new block with difficulty set to `1`.
+* If the validator is the attester of the last block, wait for `ATTEST_INTERVAL` blocks then continue this strategy.
+
+`ATTEST_INTERVAL` is used for preventing malicious validators to censors the POA network. When malicious validators are less than `ATTEST_INTERVAL + 1` that at least an honest validator have the chance to submit votes and evict malicious validators.
+
+`ATTEST_INTERVAL` can be set to `VALIDATOR_COUNT / 2` the honest validators could eventually evict malicious validators unless the half of validators are corrupted.
+
+## validator list
+
+Validator list can be represented as a list of pubkey hash.
+
+``` txt
+pubkey_hash_1 | pubkey_hash_2 | pubkey_hash_3 ...
+```
+
+We can use an `M of N` multisig lock to describe the validator list, the `M` is set to `VOTE_LIMIT`, and `N` is `VALIDATORS_COUNT`.
+
+To change the validator list, a validator can collect enough votes(the signatures) and send a tx to spent the old multisig lock and construct a new multisig lock with new pubkey hashes. A validator node can use `type_id` to track this validator list.
+
+The pre-defined [multisig script](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0021-ckb-address-format/0021-ckb-address-format.md#short-payload-format) is satisfied the requirements, one except is the `multisig script` should be public so other nodes can check new validator list. Validators should make sure the entire `multisig script` is stored to cell data, and only one cell is constructed to represent the validator list.
+
+Notice, the process of collect data is not included in the POA protocol for simplifying, a responsible validator should check the vote tx carefully before signing it.
+
+## POA header
+
+Unfortunately, CKB header is fixed-length, the `nonce` field takes 128 bits, we can't just put 256 bits secp256k1 signature into the header.
+
+For easier implementation, instead change the block header structure, we put an extra POA context concatenated after cellbase transaction's first witness. To verify a header attestation, we need both the header and the cellbase transaction.
```

Comment:
```suggestion
For the ease of implementation, instead of changing the block header structure, we put an extra POA payload `POAContext` in the cellbase transaction's first witness. To verify a header attestation, we need both the header and the cellbase transaction.
```